Ksz8863 spi bug fix i2c feature

paveldedourek-eaton opened the pull request:
- SPI communication fixed for KSZ8863
- option to communicate with KSZ8863 via I2C added
- see [Zephyr issue #42524 ](https://github.com/zephyrproject-rtos/zephyr/issues/42524)

carlescufi commented:
@arvinf and @lmajewski please take a look.

lmajewski commented on drivers/ethernet/dsa_ksz8xxx.c:
```diff
@@ -57,9 +57,15 @@ static void dsa_ksz8xxx_write_reg(const struct ksz8xxx_data *pdev,
 	buf[0] = KSZ8XXX_SPI_CMD_WR | ((reg_addr >> 7) & 0x1F);
 	buf[1] = (reg_addr << 1) & 0xFE;
 	buf[2] = value;
+  */
```
As I've said - The code which you commented out works on `ip_k66f` board. I'm able to configure the `KSZ8794` switch (setup MAC address table, get port status info, read SoC ID) - which means that the specified address is correct (at least for NXP's SPI driver).

lmajewski commented on drivers/ethernet/dsa_ksz8xxx.c:
```diff
@@ -89,9 +95,15 @@ static void dsa_ksz8xxx_read_reg(const struct ksz8xxx_data *pdev,
 	buf[2] = 0x0;
+  */
+  buf[0] = KSZ8XXX_SPI_CMD_RD;
+	buf[1] = (reg_addr ) & 0xFF;
```
For KSZ8794:
https://ww1.microchip.com/downloads/en/DeviceDoc/KSZ8794CNX-Integrated-4-Port-10-100-Managed-Ethernet-Switch-DS00002134C.pdf
3.5.1.1SPI Client Serial Bus Configuration
Figure 3-7 and 3-8

It looks like STM32 has different LSB/MSB order for the SPI.

Your change will break the already working `ip_k66f` and probably SAM (Atmel)

lmajewski commented on drivers/ethernet/dsa_ksz8xxx.c:
```diff
@@ -57,15 +57,17 @@ static void dsa_ksz8xxx_write_reg(const struct ksz8xxx_data *pdev,
-	buf[1] = (reg_addr << 1) & 0xFE;
-	buf[2] = value;
-  */
+#if CONFIG_DSA_KSZ8863
```
This needs to be merged to a single commit.

lmajewski commented on drivers/ethernet/dsa_ksz8xxx.c:
```diff
@@ -682,7 +701,7 @@ static int dsa_ksz8xxx_configure_bus(struct ksz8xxx_data *pdev)
 	pdev->spi = (struct spi_dt_spec) SPI_DT_SPEC_INST_GET(0,
 #if DT_INST_PROP(0, spi_cpol)
-		SPI_MODE_CPOL |
+		SPI_MODE_CPOL | 
```
This is probably not needed (not related to this patch series).
You can do the whitespaces "cleanup" as a separate commit.

lmajewski commented on dts/bindings/dsa/microchip_dsa-i2c.yaml:
```diff
@@ -0,0 +1,34 @@
+# Copyright (c) 2020 DENX Software Engineering GmbH
```
I think that you are the author here :-)

lmajewski reviewed (changes requested):
Thanks for your effort on providing I2C support for the DSA framework.

arvinf commented on drivers/ethernet/dsa_ksz8863.h:
```diff
@@ -8,8 +8,11 @@
+//#define KSZ8863_SPI_CMD_RD (BIT(6) | BIT(5))
+
+#define KSZ8863_SPI_CMD_WR (BIT(1))
+#define KSZ8863_SPI_CMD_RD (BIT(1) | BIT(0))
```
KSZ8863 was used with bit-bang SMI driver and not SPI but that code is not checked in. I kept the SPI support as-is but wasn't able to test it since my board didn't have SPI connection to the PHY.

lmajewski replied:
@arvinf Have I understand correctly that the SPI part of KSZ8863 was not tested at all (and just copied from KSZ8794)?

paveldedourek-eaton replied:
@lmajewski The root cause is described here: https://github.com/zephyrproject-rtos/zephyr/issues/42524 
It looks like, that KSZ8863 driver has not been test on Zephyr. It seems that KSZ8796 respect specification, but I cannot test it here, I do not have such HW.

Basically, KSZ8863 use different SPI protocol address is 8bit vs 16bit and positions of control bits are also different.

lmajewski replied:
@arvinf @paveldedourek-eaton @carlescufi 
- The SPI code for `KSZ8863` has not been tested up till @paveldedourek-eaton rework - I was just a copy & paste from KSZ8794
- The KSZ8863 datasheet:
https://ww1.microchip.com/downloads/en/DeviceDoc/KSZ8863MLL-FLL-RLL-Data-Sheet-00002335C.pdf
Figure 3-9 and Figure 3-10 shows that the addressing scheme and bit order is the same as for KSZ8794 (proper link and document reference shared in other reply in this thread).
Considering the above - the code for KSZ8794 shall work as well for KSZ8863.

Or am I missing something?

lmajewski replied:
@paveldedourek-eaton - Could you reply to the above comment? From what I can read from documentation of those two chips - there shall be no need for changing the SPI code for transmission to KSZ8863. Could you look to pointed out documentation and double check your setup? (Or correct me if I'm wrong)?

lmajewski replied:
For the record - I've pasted the PM you sent me (as you cannot add it to the github):
`SPI transaction contains in both case 3 bytes, with different content:`
`KSZ8794:`
`byte0 - 6bit+5bit -> COMMAND, 4..0bits -> address A11..A7`
`byte1 - 7..1bits -> address A6..A0, 0bit -> TR`
`byte2 - data`
`KSZ8863:`
`byte0 - 1bit+0bit -> COMMAND`
`byte1 - 7..0bits -> address A7..A0`
`byte2 - data`

As we are now 100% sure that the problem is not with LSB/MSB ordering, then IMHO we shall pull your changes.

(I'm _really_ surprised about the inconsistency about SPI transmission command byte format for both KSZ chips).

paveldedourek-eaton commented:
I am going to put it all to one commit.

paveldedourek-eaton commented:
Do you see my comments? It looks, that they are hidden. So, I will copy it here:

@lmajewski<https://github.com/lmajewski> I think, the topic regarding different SPI protocol for KSZ8794 and KSZ8863 is covered by fog.
Please, check datasheet for both devices - KSZ8794 (figure 3-7) and KSZ8893 (figure 3-11).
SPI transaction contains in both case 3 bytes, with different content:
KSZ8794:
byte0 - 6bit+5bit -> COMMAND, 4..0bits -> address A11..A7
byte1 - 7..1bits -> address A6..A0, 0bit -> TR
byte2 - data
KSZ8863:
byte0 - 1bit+0bit -> COMMAND
byte1 - 7..0bits -> address A7..A0
byte2 - data

***@***.***


Protocols are on first view the same, but in case, we have to split protocol to SPI bytes, it is different. Driver is implemented in the way where SPI is byte oriented.

Something else is missing?

Pavel

From: Łukasz Majewski ***@***.***>
Sent: Friday, February 25, 2022 10:20 PM
To: zephyrproject-rtos/zephyr ***@***.***>
Cc: Dedourek, Pavel ***@***.***>; Mention ***@***.***>
Subject: [EXTERNAL] Re: [zephyrproject-rtos/zephyr] Ksz8863 spi bug fix i2c feature (PR #43069)


@lmajewski commented on this pull request.

________________________________

In drivers/ethernet/dsa_ksz8863.h<https://github.com/zephyrproject-rtos/zephyr/pull/43069#discussion_r815162036>:

> @@ -8,8 +8,11 @@

 #define __DSA_KSZ8863_H__



 /* SPI commands */

-#define KSZ8863_SPI_CMD_WR (BIT(6))

-#define KSZ8863_SPI_CMD_RD (BIT(6) | BIT(5))

+//#define KSZ8863_SPI_CMD_WR (BIT(6))

+//#define KSZ8863_SPI_CMD_RD (BIT(6) | BIT(5))

+

+#define KSZ8863_SPI_CMD_WR (BIT(1))

+#define KSZ8863_SPI_CMD_RD (BIT(1) | BIT(0))

@paveldedourek-eaton<https://github.com/paveldedourek-eaton> - Could you reply to the above comment? From what I can read from documentation of those two chips - there shall be no need for changing the SPI code for transmission to KSZ8863. Could you look to pointed out documentation and double check your setup? (Or correct me if I'm wrong)?

—
Reply to this email directly, view it on GitHub<https://github.com/zephyrproject-rtos/zephyr/pull/43069#discussion_r815162036>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AXMZ5G3HHF52SDGFWD2R53DU47XAJANCNFSM5PBRVSDQ>.
Triage notifications on the go with GitHub Mobile for iOS<https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android<https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
You are receiving this because you were mentioned.Message ID: ***@***.******@***.***>>



________________________________
Eaton Elektrotechnika s.r.o. ~ Sídlo společnosti, jak je zapsáno v rejstříku: Komárovská 2406, Praha 9 - Horní Počernice, 193 00, Česká Republika ~ Jméno, místo, kde byla společnost zaregistrována: Praha ~ Identifikační číslo (IČO): 498 11 894

________________________________

lmajewski commented:
> Do you see my comments? It looks, that they are hidden. So, I will copy it here: @lmajewski<https://github.com/lmajewski> I think, the topic regarding different SPI protocol for KSZ8794 and KSZ8863 is covered by fog. Please, check datasheet for both devices - KSZ8794 (figure 3-7) and KSZ8893 (figure 3-11). SPI transaction contains in both case 3 bytes, with different content: KSZ8794: byte0 - 6bit+5bit -> COMMAND, 4..0bits -> address A11..A7 byte1 - 7..1bits -> address A6..A0, 0bit -> TR byte2 - data KSZ8863: byte0 - 1bit+0bit -> COMMAND byte1 - 7..0bits -> address A7..A0 byte2 - data ***@***.*** Protocols are on first view the same, but in case, we have to split protocol to SPI bytes, it is different. Driver is implemented in the way where SPI is byte oriented. Something else is missing? Pavel From: Łukasz Majewski ***@***.***> Sent: Friday, February 25, 2022 10:20 PM To: zephyrproject-rtos/zephyr ***@***.***> Cc: Dedourek, Pavel ***@***.***>; Mention ***@***.***> Subject: [EXTERNAL] Re: [zephyrproject-rtos/zephyr] Ksz8863 spi bug fix i2c feature (PR #43069) @lmajewski commented on this pull request.
> ________________________________ In drivers/ethernet/dsa_ksz8863.h<[#43069 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/43069#discussion_r815162036)>:
> @@ -8,8 +8,11 @@
> #define __DSA_KSZ8863_H__ /* SPI commands */ -#define KSZ8863_SPI_CMD_WR (BIT(6)) -#define KSZ8863_SPI_CMD_RD (BIT(6) | BIT(5)) +//#define KSZ8863_SPI_CMD_WR (BIT(6)) +//#define KSZ8863_SPI_CMD_RD (BIT(6) | BIT(5)) + +#define KSZ8863_SPI_CMD_WR (BIT(1)) +#define KSZ8863_SPI_CMD_RD (BIT(1) | BIT(0)) @paveldedourek-eaton<https://github.com/paveldedourek-eaton> - Could you reply to the above comment? From what I can read from documentation of those two chips - there shall be no need for changing the SPI code for transmission to KSZ8863. Could you look to pointed out documentation and double check your setup? (Or correct me if I'm wrong)? — Reply to this email directly, view it on GitHub<[#43069 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/43069#discussion_r815162036)>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AXMZ5G3HHF52SDGFWD2R53DU47XAJANCNFSM5PBRVSDQ>. Triage notifications on the go with GitHub Mobile for iOS<https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android<https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>. You are receiving this because you were mentioned.Message ID: ***@***.******@***.***>>
> […](#)
> ________________________________ Eaton Elektrotechnika s.r.o. ~ Sídlo společnosti, jak je zapsáno v rejstříku: Komárovská 2406, Praha 9 - Horní Počernice, 193 00, Česká Republika ~ Jméno, místo, kde byla společnost zaregistrována: Praha ~ Identifikační číslo (IČO): 498 11 894
> ________________________________

Yes, now I can see it. I've replied in the "review" section. As stated there - your patch shall be accepted.

carlescufi commented:
@paveldedourek-eaton could you please rebase and solve conflicts? Dík!

paveldedourek-eaton committed:
dsa: fix SPI for KSZ8863, I2C implemented for KSZ8863

    This is bugfix for KSZ8863 when SPI is used for
    communication. SPI protocol contains address part
    which is different between KSZ8863 and KSZ8794.
    Support for I2C communication with KSZ8863 implemented.

    Signed-off-by: Pavel Dedourek <PavelDedourek@Eaton.com>

mbolivar-nordic reviewed (commented):
Devicetree bindings changes look fine to me. I did not review the rest.

paveldedourek-eaton committed:
Merge branch 'main' into ksz8863_spi_bug_fix_i2c_feature

paveldedourek-eaton commented:
I found, there is problem in unit test:
https://github.com/zephyrproject-rtos/zephyr/runs/6106774249?check_suite_focus=true

It is about **tests/net/socket/select/** on **mps2_an385** platform. I have no idea what is wrong. I do not have mps2_an385. What can I do? How to solve this issue?

mbolivar-nordic commented:
> It is about **tests/net/socket/select/** on **mps2_an385** platform. I have no idea what is wrong. I do not have mps2_an385. What can I do? How to solve this issue?

This looks like a potential false positive due to QEMU timing issues. I would try rebasing, then ask someone to re-run CI and see if it gets fixed. Sorry about this. My understanding is that QEMU's timer code can sometimes fail tests incorrectly when the CI system running QEMU is under heavy load.

paveldedourek-eaton commented:
@mbolivar-nordic 
I can try rebase.

_I would try rebasing, then ask someone to re-run CI and see if it gets fixed._ I do not how do it... where to ask? Discord?

mbolivar-nordic commented:
> _I would try rebasing, then ask someone to re-run CI and see if it gets fixed._ I do not how do it... where to ask? Discord?

Just post a comment here; any of the reviewers ought to be able to approve the workflows.

paveldedourek-eaton commented:
Please, I would like to have the label "stale" removed. I did not have time to fully work on this topic. Currently I have to solve also problem in eth_stm32_hal.c where MII interface is used to communicate with PHY. In this case, I have to redirect it to I2C or SPI because KSZ8xxx driver does not use MII.

paveldedourek-eaton commented:
Please, I would like to have the label "stale" removed. No time to work on this topic.
